src/jdk.compiler/share/classes/com/sun/tools/javac/comp/LambdaToMethod.java

```diff
@@ -2116,11 +2116,12 @@ public Symbol baseSymbol() {
             }
 
             void addSymbol(Symbol sym, LambdaSymbolKind skind) {
-                if (skind == CAPTURED_THIS && sym != null && sym.kind == TYP && !typesUnderConstruction.isEmpty()) {
+                if (skind == CAPTURED_THIS && sym != null && sym.kind == TYP) {
                     ClassSymbol currentClass = currentClass();
-                    if (currentClass != null && typesUnderConstruction.contains(currentClass)) {
-                        // reference must be to enclosing outer instance, mutate capture kind.
-                        Assert.check(sym != currentClass); // should have been caught right in Attr
+                    if (currentClass != null && currentClass != sym && !outerThisReachable(currentClass.type, sym)) {
```

Comment:
There is a subtle change here - note that now if `currentClass == sym` we just skip this part (in this case just using `this` is enough).